drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,628 @@
+     *
+     * This value is ignored when @ref gpio_conf_t::pull is configured to
+     * @ref GPIO_FLOATING.
+     */
+    gpio_pull_strength_t pull_strength;
+    /**
+     * @brief   Drive strength of the GPIO
+     *
+     * @warning If the requested drive strength is not available, the closest
+     *          fit supported will be configured instead.
+     *
+     * This value is ignored when @ref gpio_conf_t::state is configured to
+     * @ref GPIO_INPUT or @ref GPIO_DISCONNECT.
+     */
+    gpio_drive_strength_t drive_strength;
+} gpio_conf_t;
+#endif
+
+/**
+ * @brief   A standard configuration for a generic floating input pin
+ */
+extern const gpio_conf_t gpio_ll_in;
+
+/**
+ * @brief   A standard configuration for a generic input pin with pull down
+ *          resistor
+ */
+extern const gpio_conf_t gpio_ll_in_pd;
+
+/**
+ * @brief   A standard configuration for a generic input pin with pull up
+ *          resistor
+ */
+extern const gpio_conf_t gpio_ll_in_pu;
+
+/**
+ * @brief   A standard configuration for a generic input pin with pull
+ *          resistor to keep signal at bus level
+ *
+ * This means, when the input reaches a 0, a pull down resistor is applied. If
+ * input reaches 1, a pull up is applied instead.
+ */
+extern const gpio_conf_t gpio_ll_in_pk;
+
+/**
+ * @brief   A standard configuration for a generic push-pull output pin
+ *
+ * @note    The pin will have an initial value of 0.
+ */
+extern const gpio_conf_t gpio_ll_out;
+
+/**
+ * @brief   A standard configuration for a generic floating open drain output
+ *
+ * @note    The pin will have an initial value of 1 (which in absence of an
+ *          external pull up resistor will be high impedance).
+ */
+extern const gpio_conf_t gpio_ll_od;
+
+/**
+ * @brief   A standard configuration for a generic open drain output with pull
+ *          up
+ *
+ * @note    The pin will have an initial value of 1 (so that the pull up will
+ *          pill the line high).
+ */
+extern const gpio_conf_t gpio_ll_od_pu;
+
+/**
+ * @brief   Check if the given number is a valid argument for @ref GPIO_PORT
+ *
+ * @param[in]       num     port number to check
+ * @retval          true    the MCU used has a GPIO port with that number
+ * @retval          false   the MCU used has ***NO*** GPIO port with that number
+ */
+static inline bool is_gpio_port_num_valid(uint_fast8_t num);
+
+/**
+ * @brief   Initialize the given GPIO pin as specified
+ *
+ * @param[in]       port    port the pin to initialize belongs to
+ * @param[in]       pin     number of the pin to initialize
+ * @param[in]       conf    configuration to apply
+ *
+ * @retval  0           success
+ * @retval  -ENOTSUP    GPIO state or pull resistor configuration not supported
+ *
+ * @warning If the configuration of the Schmitt trigger, the drive strength, or
+ *          the pull resistor strength are not supported, the closest supported
+ *          value will be chosen instead and `0` is returned.
```

Comment:
I'm not sure if it really is needed to specify or guideline behavior for the optional fields. Whoever implements the backend might exercise this freedom to provide a better implementation, without breaking 99% of the applications that do not really have requirements on drive strength, pull resistor strength etc. And for those who have requirements on this, the note below gives the hint to just ensure with `gpio_ll_query_conf()` that their requirements are met.

Reply:
It may easily be that we can go less prescriptive -- but the point should be made that (at least on certain platforms) the weak properties may be altered even by a subsequent pin configuration, and that thus (on all platforms as a matter of good practice), the query checks are best performed after all pins of a port have been configured.

Reply:
> but the point should be made that (at least on certain platforms) the weak properties may be altered even by a subsequent pin configuration

Just the thought of this gives me shivers. A call of `gpio_ll_init()` on pin A may limited the configuration options available to a subsequent call of `gpio_ll_init()` for pin B. But under no circumstances may a call to `gpio_ll_init()` for pin C change the configuration of an already configured pin D. Once the API has negotiated terms for a GPIO pin with the application, those terms are binding until the next call of `gpio_ll_init()` on that pin.

Reply:
Yes, it is scary, that's why it needs to be documented what happens.

Limiting future pins' options is the first in my "for example" group (which I maybe should have marked more clearly as at least the two options we have here).

Thing is, this comes with runtime cost. If the first configured pin limits the options, then the peripheral driver needs to *remember* which pins were configured (equivalently, which properties of a port have been set for a reason). As things like "alternate port driver strength" may not have a sentinel "I don't care" value, if a nonstandard strength is requested for a newly configured pin, the driver needs to go through all pins of that port, check whether they were configured in the first place (which may not be evident from the registers), check whether they use the alternative drive strength, and if none uses it, then it may configure the alternative drive strength.

The alternative is scary but way easier: Configure all pins on a port, and then verify whether all properties you really need are upheld.

My preference depends on which side of the API I'm currently at :-) -- but the API description should take a stance.

Reply:
Ah, didn't wait for the push :-)

Now the statement is there and clear, thanks. (I'll might curse later when I implement a peripheral based on it, but I promise to do so politely)

(I like the "SHOULD" in combination with the "will not be changed as a side effect": this gives LL implementers the license to take out-of-band hints, like "I have to configure some pins now but don't go out of the way using once-in-a-reset features to satisfy any detail", while still preventing hell from breaking lose)

Reply:
I expect that 99% of the interactions between pins are due to shared resources, e.g. pulling the GPIO towards bus level might be implemented by one hardware unit per GPIO port that can be routed to any but at most one pin of that port. Those likely are exposed with a single memory mapped register per port. So expectation is that the runtime cost boils down to something like this with few exceptions:

```C
    if (gpio->SOME_REG == UNUSED) {
         gpio->SOME_REG == pin;
    }
    else {
        return -ENOTSUP;
    }
```

It can be a bit more complex to manage when e.g. there is a maximum source/sink current per port given. But that won't be too bad either to manage either.

After all, `gpio_ll_init()` is not expected to be called in hot code paths, so spending a few more cycles to avoid footguns sounds like a good deal.

Also, I think it would be only fair if vendor decides to safe a bit of wafer area, that only the drivers for their hardware have to handle the fallout of this, rather than letting unsuspecting application developers handle with the fallout.